### PR TITLE
FHAC-1465: Add configurable TLS version in uddi service call

### PR DESCRIPTION
**Pull request by mhpnguyen**
- Add configurable TLS version in uddi service call only and won't affect nhinc gateway call.  If an customized entry doesn't exist gateway.propertiees, the uddi will uses jvm default option (JDK 7 uses TLSV1 while JDK8 uses TLSv1.2)
- Address some existing minor sonarqube violations.
@jsmith2 (optional)

### CODE REVIEW

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/**
+ *
+ */
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java]: Missing CONNECT license clause.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     * (non-Javadoc)
+     *
+     * @see gov.hhs.fha.nhinc.messaging.service.decorator.cxf.TLSClientServiceEndpointDecorator#configure()
+     */
+    @Override
+    public void configure() {
+        super.configure();
+        final String protocol = getSecureProtocol();
+        LOG.info("TLS support versions {}", protocol);
+        HTTPConduit conduit = getHttpConduit();
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java]: Does this value need to be stored, or can it be used inline below (e.g., getHttpConduit().setTlsClientParameters(tlsCP))?

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java [hunk 3]
```diff
@@ -64,36 +66,48 @@ private TLSClientParametersFactory() {
             trustFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
             trustFactory.init(cm.getTrustStore());
         } catch (UnrecoverableKeyException | KeyStoreException | NoSuchAlgorithmException e) {
-            LOG.error("Could not initialize key and trust managers: " + e.getLocalizedMessage(), e);
+            LOG.error("Could not initialize key and trust managers: {} ", e.getLocalizedMessage(), e);
         }
 
     }
 
     public static TLSClientParametersFactory getInstance() {
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java:72]: Consider using a Singleton pattern here.  For an example within the CONNECT codebase, see X12EntityExceptionBuilder.java.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java [hunk 4]
```diff
@@ -64,36 +66,48 @@ private TLSClientParametersFactory() {
...
     public TLSClientParameters getTLSClientParameters() {
-        TLSClientParameters tlsCP = new TLSClientParameters();
+        return constructTLSClient(new TLSClientParameters());
+    }
 
+    private static TLSClientParameters constructTLSClient(TLSClientParameters tlsClientParameters) {
         try {
-            SSLContext context = SSLContext.getDefault();
+            SSLContext context = SSLUtils.getSSLContext(tlsClientParameters);
             SSLSocketFactory factory = context.getSocketFactory();
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java]: there is no need to create context separately. You can use inline over here

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientServiceEndpointDecorator.java [hunk 5]
```diff
@@ -71,11 +71,22 @@ public TLSClientServiceEndpointDecorator(ServiceEndpoint<T> decoratoredEndpoint,
     @Override
     public void configure() {
         super.configure();
-        Client client = ClientProxy.getClient(getPort());
-        HTTPConduit conduit = (HTTPConduit) client.getConduit();
+        HTTPConduit conduit = getHttpConduit();
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientServiceEndpointDecorator.java]: Use inline method call over here.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import gov.hhs.fha.nhinc.properties.PropertyAccessException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author mpnguyen
+ *
+ */
+public class TLSUddiClientEndpointDecorator<T> extends TLSClientServiceEndpointDecorator<T> {
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java]: Within the CONNECT codebase, pick and choose whether acronyms will be camel-cased or not -- I strongly suggest not mixing and matching, e.g.:

Mixed (bad):  TLSUddi
Not mixed (okay):  TlsUddi
Not mixed (okay):  TLSUDDI

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientServiceEndpointDecorator.java [hunk 7]
```diff
@@ -71,11 +71,22 @@ public TLSClientServiceEndpointDecorator(ServiceEndpoint<T> decoratoredEndpoint,
...
+        return (HTTPConduit) client.getConduit();
+    }
+
+    /**
+     * @return the tlsClientFactory
+     */
+    public TLSClientParametersFactory getTlsClientFactory() {
+        return tlsClientFactory;
+    }
+
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientServiceEndpointDecorator.java]: remove these extra line spaces.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+     * @return
+     */
+    private String getSecureProtocol() {
+        String secureProtocol = null;
+        try {
+            secureProtocol = getPropertyAccessor().getProperty(NhincConstants.GATEWAY_PROPERTY_FILE,
+                NhincConstants.UDDI_TLS);
+            LOG.debug("Retrieve UDDI {} from {} property",secureProtocol, NhincConstants.GATEWAY_PROPERTY_FILE);
+            // convert test1;test2 into test1,test2
+            secureProtocol = StringUtils.replace(secureProtocol, ";", ",");
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java]: Is there a reason to store the values separated by semicolon

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java [hunk 9]
```diff
@@ -26,9 +26,10 @@
  */
 package gov.hhs.fha.nhinc.messaging.service.decorator.cxf;
 
+import java.security.GeneralSecurityException;
+
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java]: extra line space

**Review by TabassumJafri — CHANGES_REQUESTED** [review]
Please see my inline comments

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java (hunk 8 above)
Comment (mhpnguyen) [reply Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSUddiClientEndpointDecorator.java]: PropertyFileDAO and other places only allows ";".  To change this feature, we need to modify several places which are doable but they are outside of this scope.  We can create new ticket to track this issue

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java (hunk 3 above)
Comment (mhpnguyen) [reply Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/cxf/TLSClientParametersFactory.java:72]: Done. Convert this to follow Singleton Pattern.

**Review by TabassumJafri — APPROVED** [review]

**Review by balmeeki — APPROVED** [review]

**Review by TabassumJafri — APPROVED** [review]
Good with merge

**Review by jsmith2 — APPROVED** [review]

**Pull request merged**